Backports for julia 1.11.7

KristofferC opened the pull request:
Backported PRs:
- [x] #54840 <!-- Add boundscheck in speccache_eq to avoid OOB access due to data race -->
- [x] #42080 <!-- recommend explicit `using Foo: Foo, ...` in package code (was: "using considered harmful") -->
- [x] #58127 <!-- [DOC] Update installation docs: /downloads/ => /install/ -->
- [x] #58202 <!-- [release-1.11] malloc: use jl_get_current_task to fix null check -->
- [x] #58584 <!-- Make `Ptr` values static-show w/ type-information -->
- [x] #58637 <!-- Make late gc lower handle insertelement of alloca use. -->
- [x] #58837 <!-- fix null comparisons for non-standard address spaces -->
- [x] #57826 <!-- Add a `similar` method for `Type{<:CodeUnits}` -->
- [x] #58293 <!-- fix trailing indices stackoverflow in reinterpreted array -->
- [x] #58887 <!-- Pkg: Allow configuring can_fancyprint(io::IO) using IOContext -->
- [x] #58937 <!-- Fix nthreadpools size in JLOptions -->
- [x] #58978 <!-- Fix precompilepkgs warn loaded setting -->
- [x] #58998 <!-- Bugfix: Use Base.aligned_sizeof instead of sizeof in Mmap.mmap -->
- [x] #59120 <!-- Fix memory order typo in "src/julia_atomics.h" -->
- [x] #59170 <!-- Clarify and enhance confusing precompile test -->

Need manual backport:
- [ ] #56329 <!-- loading: clean up more concurrency issues -->
- [ ] #56956 <!-- Add "mea culpa" to foreign module assignment error. -->
- [ ] #57035 <!-- linux: workaround to avoid deadlock inside dl_iterate_phdr in glibc -->
- [ ] #57089 <!-- Block thread from receiving profile signal with stackwalk lock -->
- [ ] #57249 <!-- restore non-freebsd-unix fix for profiling -->
- [ ] #58011 <!-- Remove try-finally scope from `@time_imports` `@trace_compile` `@trace_dispatch` -->
- [ ] #58062 <!-- remove unnecessary edge from `exp_impl` to `pow` -->
- [ ] #58157 <!-- add showing a string to REPL precompile workload -->
- [ ] #58209 <!-- Specialize `one` for the `SizedArray` test helper -->
- [ ] #58108 <!-- Base.get_extension & Dates.format made public -->
- [ ] #58356 <!-- codegen: remove readonly from abstract type calling convention -->
- [ ] #58415 <!-- [REPL] more reliable extension loading -->
- [ ] #58510 <!-- Don't filter `Core` methods from newly-inferred list -->
- [ ] #58110 <!-- relax dispatch for the `IteratorSize` method for `Generator` -->
- [ ] #58965 <!-- Fix `hygienic-scope`s in inner macro expansions -->
- [ ] #58971 <!-- Fix alignment of failed precompile jobs on CI -->
- [ ] #59066 <!-- build: Also pass -fno-strict-aliasing for C++ -->

Contains multiple commits, manual intervention needed:
- [ ] #55877 <!-- fix FileWatching designs and add workaround for a stat bug on Apple -->
- [ ] #56755 <!-- docs: fix scope type of a `struct` to hard -->
- [ ] #57809 <!-- Fix fptrunc Float64 -> Float16 rounding through Float32 -->
- [ ] #57398 <!-- Make remaining float intrinsics require float arguments -->
- [ ] #56351 <!-- Fix `--project=@script` when outside script directory -->
- [ ] #57129 <!-- clarify that time_ns is monotonic -->
- [ ] #58134 <!-- Note annotated string API is experimental in Julia 1.11 in HISTORY.md -->
- [ ] #58401 <!-- check that hashing of types does not foreigncall (`jl_type_hash` is concrete evaluated) -->
- [ ] #58435 <!-- Fix layout flags for types that have oddly sized primitive type fields -->
- [ ] #58483 <!-- Fix tbaa usage when storing into heap allocated immutable structs -->
- [ ] #58512 <!-- Make more types jl_static_show readably -->
- [ ] #58012 <!-- Re-enable tab completion of kwargs for large method tables -->
- [ ] #58683 <!-- Add 0 predecessor to entry basic block and handle it in inlining -->
- [ ] #59112 <!-- Add builtin function name to add methods error -->

Non-merged PRs with backport label:
- [ ] #59329 <!-- aotcompile: destroy LLVM context after serializing combined module -->
- [ ] #58848 <!-- Set array size only when safe to do so -->
- [ ] #58535 <!-- gf.c: include const-return methods in `--trace-compile` -->
- [ ] #58038 <!-- strings/cstring: `transcode`: prevent Windows sysimage invalidation -->
- [ ] #57604 <!-- `@nospecialize` for `string_index_err` -->
- [ ] #57366 <!-- Use ptrdiff_t sized offsets for gvars_offsets to allow large sysimages -->
- [ ] #56890 <!-- Enable getting non-boxed LLVM type from Julia Type -->
- [ ] #56823 <!-- Make version of opaque closure constructor in world -->
- [ ] #55958 <!-- also redirect JL_STDERR etc. when redirecting to devnull -->
- [ ] #55956 <!-- Make threadcall gc safe -->
- [ ] #55534 <!-- Set stdlib sources as read-only during installation -->
- [ ] #55499 <!-- propagate the terminal's `displaysize` to the `IOContext` used by the REPL -->
- [ ] #55458 <!-- Allow for generically extracting unannotated string -->
- [ ] #55457 <!-- Make AnnotateChar equality consider annotations -->
- [ ] #55220 <!-- `isfile_casesensitive` fixes on Windows -->
- [ ] #53957 <!-- tweak how filtering is done for what packages should be precompiled -->
- [ ] #51479 <!-- prevent code loading from lookin in the versioned environment when building Julia -->
- [ ] #50813 <!-- More doctests for Sockets and capitalization fix -->
- [ ] #50157 <!-- improve docs for `@inbounds` and `Base.@propagate_inbounds` -->

kpamnany committed:
Add boundscheck in speccache_eq to avoid OOB access due to data race (#54840)

Like https://github.com/JuliaLang/julia/pull/54671, but for
`speccache_eq`.

Saw another segfault with this in the stack trace, hence this fix. I
also looked for other uses of `jl_smallintset_lookup` and there's one in
`idset.c`. That doesn't appear to be racy but I'm not familiar with the
code, so maybe you can take a look at it in case we need to push a fix
for that one too @gbaraldi or @vtjnash?

(cherry picked from commit dd1ed17ae3c2e4e257f2444bbafd705326b4bbbd)

adienes committed:
Switch from segfault to `zip` behavior for mismatched indices in `map!` (#56673)

(cherry picked from commit 0947114d9d443e14c751ac40c9b2d8c2245d045e)

gbaraldi committed:
Make late gc lower handle insertelement of alloca use. (#58637)

This was in DAECompiler.jl code found by @serenity4. He also mentioned
that writing up how one might go and fix a bug like this so i'll give a
quick writeup (this was a very simple bug so it might not be too
interesting)

The original crash which looked something like
>   %19 = alloca [10 x i64], align 8
  %155 = insertelement <4 x ptr> poison, ptr %19, i32 0
Unexpected instruction
> [898844] signal 6 (-6): Aborted
in expression starting at
/home/gbaraldi/DAECompiler.jl/test/reflection.jl:28
pthread_kill at /lib/x86_64-linux-gnu/libc.so.6 (unknown line)
gsignal at /lib/x86_64-linux-gnu/libc.so.6 (unknown line)
abort at /lib/x86_64-linux-gnu/libc.so.6 (unknown line)
RecursivelyVisit<llvm::IntrinsicInst,
LateLowerGCFrame::PlaceRootsAndUpdateCalls(llvm::ArrayRef<int>, int,
State&, std::map<llvm::Value*, std::pair<int, int>
>)::<lambda(llvm::AllocaInst*&)>::<lambda(llvm::Use&)> > at
/home/gbaraldi/julia4/src/llvm-late-gc-lowering.cpp:803
operator() at /home/gbaraldi/julia4/src/llvm-late-gc-lowering.cpp:2560
[inlined]
PlaceRootsAndUpdateCalls at
/home/gbaraldi/julia4/src/llvm-late-gc-lowering.cpp:2576
runOnFunction at
/home/gbaraldi/julia4/src/llvm-late-gc-lowering.cpp:2638
run at /home/gbaraldi/julia4/src/llvm-late-gc-lowering.cpp:2675
run at
/home/gbaraldi/julia4/usr/include/llvm/IR/PassManagerInternal.h:91

which means it was crashing inside of late-gc-lowering, so the first
thing I did was ran julia and the same test with LLVM_ASSERTIONS=1 and
FORCE_ASSERTIONS=1 to see if LLVM complained about a malformed module,
and both were fine. Next step was trying to get the failing code out for
inspection.
Easiest way is to do `export
JULIA_LLVM_ARGS="--print-before=LateLowerGCFrame --print-module-scope"`
and pipe the output to a file.
The file is huge, but since it's a crash in LLVM we know that the last
thing is what we want, and that gave me the IR I wanted.
To verify that this is failing I did `make -C src install-analysis-deps`
to install the LLVM machinery (opt...). That gets put in the `tools`
directory of a julia build. Then I checked if this crashed outside of
julia by doing
`./opt -load-pass-plugin=../lib/libjulia-codegen.dylib
--passes=LateLowerGCFrame -S test.ll -o tmp3.ll `. This is run from
inside the tools dir so your paths might vary (the -S is so LLVM doesn't
generate bitcode) and my code did crash, however it was over 500 lines
of IR which makes it harder to debug and to write a test.

Next step then is to minimize the crash by doing
[`llvm-reduce`](https://llvm.org/docs/CommandGuide/llvm-reduce.html)
over it (it's basically creduce but optimized for LLVM IR) which gave me
a 2 line reproducer (in this case apparently just having the
insertelement was enough for the pass to fail). One thing to be wary is
that llvm-reduce will usually make very weird code, so it might be
useful to modify the code slightly so it doesn't look odd (it will have
unreachable basic-blocks and such).
After the cleanup fixing the bug here wasn't interesting but this
doesn't apply generally. And also always transform your reduced IR into
a test to put in llvmpasses.

(cherry picked from commit 906d3482d318868e7d3eef6b1db97ef5bedd6f82)

Keno committed:
Unicode: Force-inline isgraphemebreak! (#58674)

When this API was added, this function inlined, which is important,
because the API relies on the allocation of the `Ref` being elided. At
some point (I went back to 1.8) this regressed. For example, it is
currently responsible for substantially all non-Expr allocations in
JuliaParser. Before (parsing all of Base with JuliaParser):
```
│     Memory estimate: 76.93 MiB, allocs estimate: 719922.
```
After:
```
│     Memory estimate: 53.31 MiB, allocs estimate: 156.
```

Also add a test to make sure this doesn't regress again.

(cherry picked from commit d6294ba973db1dea9dc932779008fd66d27c4bd2)

simeonschaub committed:
fix null comparisons for non-standard address spaces (#58837)

Co-authored-by: Jameson Nash <vtjnash@gmail.com>
(cherry picked from commit 3ed13ea7ab3d73f408b12a70ad29565c97bf5562)

ararslan committed:
Add a `similar` method for `Type{<:CodeUnits}` (#57826)

Currently, `similar(::CodeUnits)` works as expected by going through the
generic `AbstractArray` method. However, the fallback method hit by
`similar(::Type{<:CodeUnits}, dims)` does not work, as it assumes the
existence of a constructor that accepts an `UndefInitializer`. This can
be made to work by defining a corresponding `similar` method that
returns an `Array`.

One could make a case that this is a bugfix since it was arguably a bug
that this method didn't work given that `CodeUnits` is an
`AbstractArray` subtype and the other `similar` methods work. If anybody
buys that argument, it could be nice to backport this; it came up in
some internal code that uses Arrow.jl and JSON3.jl together.

(cherry picked from commit 8e524c73804a9615dd68011b2c5741947d19bbb6)

adienes committed:
fix trailing indices stackoverflow in reinterpreted array (#58293)

would fix https://github.com/JuliaLang/julia/issues/57170, fix
https://github.com/JuliaLang/julia/issues/54623

@nanosoldier `runbenchmarks("array", vs=":master")`

(cherry picked from commit e853a4f8bac99c6458cf3e09d95d9bea67b687e7)

fonsp committed:
Pkg: Allow configuring can_fancyprint(io::IO) using IOContext (#58887)

(cherry picked from commit c0cc1e1022b780a3b7de6ad33593ef12484e46e2)

IanButterworth committed:
Fix nthreadpools size in JLOptions (#58937)

(cherry picked from commit 2349f431b1d0f97f3f06f8adc5d335b89d5f061c)

IanButterworth committed:
Fix precompilepkgs warn loaded setting (#58978)

(cherry picked from commit 107e1acf5fe79e99a39a57b2aae7045bb5bfcf72)

JonasIsensee committed:
Bugfix: Use Base.aligned_sizeof instead of sizeof in Mmap.mmap (#58998)

fix #58982

(cherry picked from commit c30199fe17aa24f77530bd64f7762546c15e2941)

omus committed:
Fix memory order typo in "src/julia_atomics.h" (#59120)

(cherry picked from commit fde79f5bb5c4835189bfe3215521137176ac3eb8)

IanButterworth committed:
Clarify and enhance confusing precompile test (#59170)

(cherry picked from commit 1f6eff183db0b947632f780b4acc440d9c41a6e2)

DilumAluthge committed:
[backports-release-1.11] Change Distributed branch from `master` to `release-julia-1.11` (but keep the commit the same) (#59390)

This PR will be followed up with a stdlib bump with the Distributed
backports.

DilumAluthgeBot committed:
🤖 [backports-release-1.11] Bump the Distributed stdlib from 6c7cdb5 to e9b9023 (#59391)

Stdlib: Distributed
URL: https://github.com/JuliaLang/Distributed.jl
Stdlib branch: release-julia-1.11
Julia branch: backports-release-1.11
Old commit: 6c7cdb5
New commit: e9b9023
Julia version: 1.11.5
Distributed version: 1.11.0(Does not match)
Bump invoked by: @DilumAluthge
Powered by:
[BumpStdlibs.jl](https://github.com/JuliaLang/BumpStdlibs.jl)

Diff:
https://github.com/JuliaLang/Distributed.jl/compare/6c7cdb5860fa5cb9ca191ce9c52a3d25a9ab3781...e9b902372d509cfc11d3c85344acbb9c9e4fa5fe

```
$ git log --oneline 6c7cdb5..e9b9023
e9b9023 Backports for Julia 1.11.x (#141)
```

Co-authored-by: DilumAluthge <5619885+DilumAluthge@users.noreply.github.com>

DilumAluthge commented:
Okay, I've pushed two commits to this branch:
1. https://github.com/JuliaLang/julia/commit/5225013a8bff7e14a78738ccaa2fcdc9f8b0eb98 (https://github.com/JuliaLang/julia/pull/59390)
2. https://github.com/JuliaLang/julia/commit/a5f88ec17664a6f443b9a64f9d4c720f578cf507 (https://github.com/JuliaLang/julia/pull/59391)

@KristofferC Should I edit the PR description manually and add these two PRs to the list? Or would that break the backporter script?

DilumAluthge committed:
Revert "Unicode: Force-inline isgraphemebreak! (#58674)"

This reverts commit 47ebf952ea2bc46d9d27a0668edff23b0956a35d.

DilumAluthge committed:
Merge branch 'release-1.11' into backports-release-1.11

DilumAluthge commented:
CI is currently failing on this PR, because of segfaults in multiple test sets.

Here is a MWE that reproduces the segfault:

```
test_list = [
    "ambiguous",
]

Base.runtests(test_list)
```

With that MWE, bisect blames 3b04664577713059bee663fb3d1cd37080a7ff8e (which backports #58837):

```
3b04664577713059bee663fb3d1cd37080a7ff8e is the first bad commit
commit 3b04664577713059bee663fb3d1cd37080a7ff8e
Author: Simeon David Schaub <simeon@schaub.rocks>
Date:   Tue Jul 1 16:58:39 2025 +0200

    fix null comparisons for non-standard address spaces (#58837)

    Co-authored-by: Jameson Nash <vtjnash@gmail.com>
    (cherry picked from commit 3ed13ea7ab3d73f408b12a70ad29565c97bf5562)

 src/cgutils.cpp | 9 ++++++++-
 1 file changed, 8 insertions(+), 1 deletion(-)
bisect found first bad commit
```

<details>

<summary>
Here are the details of the bisect (click to expand):
</summary>

I ran the bisect on Linux x86_64.

My starting good commit was: 9615af0f

My starting bad commit was: 562699a0

Here is my `git bisect run` script (`../bisect_script.bash`):

```bash
#!/usr/bin/env bash

set -euf -o pipefail

which -a ccache

make cleanall
make distclean

rm -f Make.user
echo "USECCACHE=1" >> Make.user

make -j16

./julia ../test.jl
```

Here are the contents of the `../test.jl` file:

```julia
test_list = [
    "ambiguous",
]

Base.runtests(test_list)
```

</details>

DilumAluthge commented:
I've confirmed that reverting 3b04664577713059bee663fb3d1cd37080a7ff8e locally fixes my MWE.

DilumAluthge committed:
[`backports-release-1.11`]: Revert "fix null comparisons for non-standard address spaces (#58837)" (#59445)

This reverts commit 3b04664577713059bee663fb3d1cd37080a7ff8e (which
backports #58837 to 1.11.x).

For more details, see
https://github.com/JuliaLang/julia/pull/59336#issuecomment-3239568360
and
https://github.com/JuliaLang/julia/pull/59336#issuecomment-3239575723.

Targets `backports-release-1.11` (#59336).

DilumAluthge commented:
Okay, so the `llvmpasses` CI job is currently failing on this PR.

The steps to reproduce the failure are:

```bash
make -j16 release JULIA_PRECOMPILE=0

make -j16 -C src install-analysis-deps

# Run the llvmpasses tests:
make -C test/llvmpasses
```

I've bisected the failure to 47d905df994c69f91ac416d72ec0481c1fd011da (which backports PRNUMBERTODO):

```
47d905df994c69f91ac416d72ec0481c1fd011da is the first bad commit
commit 47d905df994c69f91ac416d72ec0481c1fd011da
Author: Gabriel Baraldi <baraldigabriel@gmail.com>
Date:   Wed Jun 4 20:30:23 2025 -0300

    Make late gc lower handle insertelement of alloca use. (#58637)
```

<details>

<summary>
Here are the details of the bisect (click to expand):
</summary>

I ran the bisect on Linux x86_64.

My starting good commit was: TODO

My starting bad commit was: TODO

Here is my `git bisect run` script:

```bash
#!/bin/bash

set -euf -o pipefail

which -a ccache

make cleanall
make distclean

rm -f Make.user
echo "USE_CCACHE=1" > Make.user

make -j16 release JULIA_PRECOMPILE=0

make -j16 -C src install-analysis-deps

# Run the llvmpasses tests:
make -C test/llvmpasses
```

The `git bisect log` is:

```
[daluthge@cc-linux3 julia]$ git bisect log
git bisect start
# status: waiting for both good and bad commits
# good: [9615af0f269df4d371b8010e9507ed5bae86103b] set VERSION to 1.11.6 (#58935)
git bisect good 9615af0f269df4d371b8010e9507ed5bae86103b
# status: waiting for bad commit, 1 good commit known
# bad: [ddcbe4dd84c55dbbe2374b673aad13c340f09b8f] [`backports-release-1.11`]: Revert "fix null comparisons for non-standard address spaces (#58837)" (#59445)
git bisect bad ddcbe4dd84c55dbbe2374b673aad13c340f09b8f
# bad: [69bfb6b2daaadad732123e98cce35b146e01fc50] Fix nthreadpools size in JLOptions (#58937)
git bisect bad 69bfb6b2daaadad732123e98cce35b146e01fc50
# bad: [47ebf952ea2bc46d9d27a0668edff23b0956a35d] Unicode: Force-inline isgraphemebreak! (#58674)
git bisect bad 47ebf952ea2bc46d9d27a0668edff23b0956a35d
# good: [5361f7c3ba55baa4f4bb1b0effa4092d2cf92469] Switch from segfault to `zip` behavior for mismatched indices in `map!` (#56673)
git bisect good 5361f7c3ba55baa4f4bb1b0effa4092d2cf92469
# bad: [47d905df994c69f91ac416d72ec0481c1fd011da] Make late gc lower handle insertelement of alloca use. (#58637)
git bisect bad 47d905df994c69f91ac416d72ec0481c1fd011da
# first bad commit: [47d905df994c69f91ac416d72ec0481c1fd011da] Make late gc lower handle insertelement of alloca use. (#58637)
```

</details>

DilumAluthge committed:
("remove a testset from MMAP that might cause CI to now fail on Windows") (#59452)

DilumAluthge commented:
Okay, we are very close to green CI on this PR now. The last failure is `llvmpasses`, which would be fixed by https://github.com/JuliaRegistries/General/pull/136859/commits/b980876d66c01551d4318d048ab37545aaef7386

gbaraldi committed:
Remove failing test

DilumAluthge commented:
Alright, I'll wait for CI to finish (just to double-check that it is all green now), and then I'll kick off PkgEval.

DilumAluthge commented:
Much of CI is green now, including the previously-failing `llvmpasses` job. I'll go ahead and kick off PkgEval now, while we wait for the rest of CI to finish.

DilumAluthge commented:
CI is all green now.

KristofferC commented:
Hm, I don't see the nanosoldier notification showing that it is running.

maleadt commented:
@nanosoldier `runtests()`

DilumAluthge commented:
@nanosoldier `runtests(["FastPower", "JuliaInterpreter", "MatrixBandwidth", "ColPack", "Intervals", "MRICoilSensitivities", "FindMinimaxPolynomial", "Ferrite", "Clarabel", "MathOptChordalDecomposition", "FourierTools", "KSVD", "StateSpaceDynamics", "LaplacianExpectationMaximization", "InferOpt", "NonconvexMultistart", "QuantumSymbolics", "BetaML", "JudiLing", "ONSAS", "ParametrisedConvexApproximators", "GeoStatsValidation", "Knockoffs", "ModelingToolkitTolerances", "SurfaceReactions", "SurfaceCoverage", "AcousticRayTracers", "SmoothPeriodicStatsModels", "CalibrateEmulateSample", "StateSpaceAnalysis", "ControlBarrierFunctions", "FourLeafMLE", "RobustBlindVerification", "LinearSolveAutotune"])`

DilumAluthge commented:
@KristofferC The JuliaInterpreter failure looks real?

```
Check builtin.jl consistency: Error During Test at /home/pkgeval/.julia/packages/JuliaInterpreter/378J1/test/check_builtins.jl:6
  Got exception outside of a @test
  LoadError: UndefVarError: `isdefinedglobal` not defined in `Main`
  Stacktrace:
```

KristofferC commented:
Nah, it just that test didn't run at all on the baseline due to:

https://github.com/JuliaDebug/JuliaInterpreter.jl/blob/da8883712ddcf1e781894a9cdd96578a9f991113/test/check_builtins.jl#L4

DilumAluthge commented:
I'll try it again, just to make sure it still reproduces.

DilumAluthge commented:
@nanosoldier `runtests(["ONSAS"])`

DilumAluthge commented:
To summarize the current status of PkgEval on this PR:

There are currently two packages failing:
1. JuliaInterpreter
    - Per Kristoffer's explanation above, we can ignore this one.
2. ONSAS
    - Per what I wrote above, this failure is reproducible (although it is non-deterministic) on the tip of `release-1.11`. Therefore, I don't think this failure was introduced by a commit in this backports PR.

Therefore, my conclusion is that PkgEval looks good on this PR.

DilumAluthge commented:
I've fixed this. It's now a regular merge commit (not a squash merge). This will allow people to bisect on the `release-1.11` branch.
